fix(service-worker): continue serving api requests on cache failure

apocalyp0sys opened the pull request:
## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix

## What is the current behavior?
When responses are cached ok during sw initialization, but caching throws an error when handling api response, this response never gets to client. 

Issue Number: #21412 

## What is the new behavior?
Fix response delivery by catching errors and add 2 test cases.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

gkalpak commented on packages/service-worker/worker/src/data.ts:
```diff
@@ -432,10 +432,11 @@ export class DataGroup {
+      await this.cacheResponse(req, await res, lru, okToCacheOpaque);
     } catch {
       // TODO: handle this error somehow?
     }
```
I think it makes sense to log a warning here, when updating the cache fails (but not when the request fails). I would change the method body to something like:

```ts
private async safeCacheResponse(
    req: Request, resOrPromise: Response|Promise<Response>, lru: LruList,
    okToCacheOpaque?: boolean): Promise<void> {
  try {
    const res = await resOrPromise;
    try {
      await this.cacheResponse(req, res, lru, okToCacheOpaque);
    } catch (err) {
      // Saving the API response failed. This could be a result of a full storage.
      // Since this data is cached lazily and temporarily, continue serving clients as usual.
      this.debugHandler.log(
          err,
          `DataGroup(${this.config.name}@${this.config.version}).safeCacheResponse(${req.url}, status: ${res.status})`);

      // TODO: Better detect/handle full storage; e.g. using
      // [navigator.storage](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorStorage/storage).
    }
  } catch {
    // TODO: Handle this error somehow?
  }
}
```

This would require passing the [DebugHandler][1] from the `Driver` down to the `DataGroup`s.

[1]: https://github.com/angular/angular/blob/64823f561c2c25f05e9bce04de68374ce997548a/packages/service-worker/worker/src/driver.ts#L108

apocalyp0sys replied:
Used this code, thanks!

gkalpak reviewed (changes requested):
Thx, @apocalyp0sys! This goes into the right direction. (Also, love the tests 😍)
I have left a couple of suggestions. In addition, could you split up the PR into two commits: a `refactor` one to add the `lru` parameter to `safeCacheResponse` (since that is actually unrelated to the fix) and a `fix` one to actually fix the bug?

apocalyp0sys commented:
Hi, @gkalpak! Thanks for review.

So, I should open a new `refactor` PR, wait fot it to be merged, and then modify this PR? Or they should be completely independent?

apocalyp0sys committed:
refactor(service-worker): make signatures of caching methods compatible

Make safe caching and unsafe caching methods compatible so they can be
swapped. Gives more flexibility when writing http response processing
code.

gkalpak commented on packages/service-worker/worker/src/app-version.ts:
```diff
@@ -60,7 +61,8 @@ export class AppVersion implements UpdateSource {
       private scope: ServiceWorkerGlobalScope, private adapter: Adapter, private database: Database,
-      private idle: IdleScheduler, readonly manifest: Manifest, readonly manifestHash: string) {
+      private idle: IdleScheduler, readonly manifest: Manifest, readonly manifestHash: string,
+      private debugHandler: DebugHandler) {
```
Nit: could you move the `debugHandler` argument right after `idle` (to group "utilities/helpers" and "data" separately)?

gkalpak commented on packages/service-worker/worker/src/data.ts:
```diff
@@ -243,7 +244,8 @@ export class DataGroup {
       private scope: ServiceWorkerGlobalScope, private adapter: Adapter,
-      private config: DataGroupConfig, private db: Database, private prefix: string) {
+      private config: DataGroupConfig, private db: Database, private prefix: string,
+      private debugHandler: DebugHandler) {
```
Nit: could you move the `debugHandler` argument right after `db` (to group "utilities/helpers" and "data" separately)?

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1230,6 +1246,36 @@ import {SwTestHarness, SwTestHarnessBuilder} from '../testing/scope';
+           expect(await makeRequest(scope, '/api/foo')).toEqual('this is api foo');
+           expect(driver.state).toBe(DriverReadyState.NORMAL);
+           server.assertSawRequestFor('/api/foo');
+         });
```
Could you also add an assertion that an error was logged with `DebugHandler` (here and below)?
(I know we usually don't do that in other tests, but I think it is a good idea, especially here that we basically ignore the error.)

apocalyp0sys replied:
Should I create `MockDebugHandler` or spying on existing is ok?

apocalyp0sys replied:
Implemented spying on existing `DebugHandler`. Feel free to start another conversation if anything is not ok.

gkalpak replied:
Spying on the existing one sounds good. Thx!

gkalpak reviewed (changes requested):
Awesome! Thx for making the changes, @apocalyp0sys 👍
I left a couple more minor comments, but otherwise lgtm :tada:

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -6,6 +6,8 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
+import {DebugHandler} from '@angular/service-worker/worker/src/debug';
```
This should be a relative import, like the rest.
(Although you might not need this at all - see below.)

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1182,6 +1200,9 @@ import {SwTestHarness, SwTestHarnessBuilder} from '../testing/scope';
 
     describe('bugs', () => {
+
+      const cacheErrorMessage = 'Can\'t touch this';
```
Good thought :smiley: However, I think this is better inlined, since the message itself doesn't have any significance (it could be any message) and it makes the code more readable when inlined (which is far more important that performance or size in testing code).

(If anything, you could define a local var inside each test.)

gkalpak commented on packages/service-worker/worker/test/happy_spec.ts:
```diff
@@ -1222,14 +1243,56 @@ import {SwTestHarness, SwTestHarnessBuilder} from '../testing/scope';
+
+           // Make the caches unwritable.
+           spyOn(MockCache.prototype, 'put').and.throwError(cacheErrorMessage);
+           spyOn(DebugHandler.prototype, 'log');
```
Could you spy on `driver.debugger` instead to make sure that the same `DebugHandler` is used (and not for example a different instance).

gkalpak reviewed (changes requested):
Yay! Almost there 😅

apocalyp0sys committed:
fix(service-worker): continue serving api requests on cache failure

When responses are cached ok during sw initialization,
but caching throws an error when handling api response,
this response never gets to client. Fix response
delivery by catching errors, add logging and 2 test cases.

Fixes #21412

gkalpak reviewed (approved):
LGTM :tada: Thx, @apocalyp0sys! ❤️

mhevery commented:
I was unable to merge this to the patch branch as it causes conflicts. Could we create a new PR for patch branch?

apocalyp0sys commented:
Should i create a new PR targeting `8.2.x` branch?

gkalpak commented:
Yes, please, @apocalyp0sys 🙏
